Add support for 8-bit ANSI color codes

OlivierSohn opened the pull request:
Closes #37

feuerbach commented on System/Console/ANSI/Types.hs:
```diff
@@ -26,11 +30,24 @@ data Color = Black
                     deriving (Eq, Ord, Bounded, Enum, Show, Read, Ix)
 
+
+-- | ANSI 8-bit "6 × 6 × 6 cube" rgb colors (216 colors). Each individual vomponent values is in range [0..5].
```
typo: "vomponent". Also, should "values" be singular instead?

OlivierSohn replied:
Thanks, I will fix it!

feuerbach commented:
@mpilgrem could you take a look?

OlivierSohn commented:
I added some doc, and removed the use if Int8: I used it first thinking it was a smaller type than Int but actually under the hood it's an Int so there is no benefit memory-wise, and type-wise it doesn't make much sense either, as what we really would like to express is that the values should be between 0 and 5 for rgb or 0 and 23 for grayscale, so it doesn't help.

mpilgrem commented:
I think something needs to be added to `System\Console\ANSI\Windows\Emulator.hs` to deal with the missing pattern matches - some rational mapping of the 8-bit colours onto the available 3/4-bit colours.

OlivierSohn commented:
Oups, so the build on windows fails?

I'm not familiar with that part of the code, can someone look into it?

I'm also considering adding another pattern match will will be handy for my application, where we pass the code of the 8bit color. But it could be a separate pull request if we want to keep the complexity minimal here...

mpilgrem commented:
I could look at the emulation code. A first step is to understand how xterm maps 8-bit colours on to 24-bit colours, as modern ANSI-capable Windows follows xterm as regards 8-bit. Once you have the 24-bit, there is already code that maps that to 3/4 bit for emulation purposes on legacy Windows.

mpilgrem commented:
This resource appears to answer the first step: https://jonasjacek.github.io/colors/ The rest should be straight-forward.

OlivierSohn commented:
ok, that should be straightforward, as the formula for rgb 8bit is (from https://en.wikipedia.org/wiki/ANSI_escape_code):

> 0x10-0xE7:  6 × 6 × 6 cube (216 colors): 16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5)

we can deduce r g b from the color code.

But it's even easier because since in the pattern match we have the r g and b values, we just need to map the range [0..5] to the range for 24 bits [0..255] (I guess a linear mapping would be ok?).
For the grayscale case, it's the same thing except the source range is [0..23] and target is |0..255]

I don't know how xtern does it specifically but I guess that is a good enough first step?

OlivierSohn commented:
I was trying to compile on windows but my windows laptop just passed away... One thing I can do is write the conversion functions : 
```haskell
:: RGB8Color -> Colour Float
:: Gray8Color -> Colour Float
```

OlivierSohn commented:
I added the conversion functions in a new module, using the conversion functions of Data.Colour.SRGB, hope it helps!

OlivierSohn commented:
But actually colorspace-wise it might be wrong. I wonder if the values given in https://jonasjacek.github.io/colors/ are already in the srgb space? And in what space are we supposed to pass the values of ANSI color to the terminal?

OlivierSohn commented:
In the last commit "Fix colorspace" I use linear SRGB, I think it's what we want here (I assumed that the values we see both in https://jonasjacek.github.io/colors/ and in https://en.wikipedia.org/wiki/ANSI_escape_code don't need firther colorspace conversion before being passed as escape sequence).

mpilgrem commented:
@OlivierSohn, comparing the Wikipedia article and the 'cheat sheet', if they are both reliable, it appears possible that the first 16 8-bit colours may not necessarily duplicate the 16 3/4 bit colours. I have not yet tested that empirically. If that proves correct, it may be useful to have a data constructor for the first 16 8-bit colours. That leads me to a second thought: perhaps `RBG8Color`, `Gray8Color` etc should be alternative constructors for a single type (`Color8`?) rather than single constructors for separate types.

OlivierSohn commented:
@mpilgrem I assumed they matched, the column of "CMD" terminal matched the numbers in the cheat sheet. 

I agree on the idea of having a Color8 type, I'll change that.

OlivierSohn commented:
@mpilgrem PTAL, I also added a Color8Code type which allows to pass the raw 8bit color code (which is exactly the API I need in my application : it also allows to store colors in a memory efficient way!)

I agree that there could be a need to add the 16 missing constructors to have full 8-bit support but if you don't mind I'll just open an issue about that to be handeled later (as of today we can have the same functionality using the existing Color and ColorIntensity, or use Color8Code directly).

OlivierSohn commented:
ok, my PC decided to resurrect so I did the changes for windows terminal emulation which I'll push soon

mpilgrem commented:
@OlivierSohn, to clear off my first thought, I have tested, on ANSI-capable Windows, that the first 16 8-bit colours are the same as the 16 3/4-bit colours (the 'cheat sheet' is ambiguous in that regard, the first 16 are indeed 'system' colours but the system colours are not necessarily the 24-bit RGB values listed in the sheet).

I know from past conversations with @Feuerbach that he will want you, in due course, to squash your various commits into a single logical commit.

mpilgrem commented on System/Console/ANSI/Color.hs:
```diff
@@ -0,0 +1,102 @@
+{-# INLINE xtermMapRGB8bitComponent #-}
+xtermMapRGB8bitComponent :: Int -> Int
+xtermMapRGB8bitComponent 0 = 0
+xtermMapRGB8bitComponent n = 95 + (pred n) * 40
```
Perhaps expressed more simply as the equivalent:
`xtermMapRGB8bitComponent n = 55 + n * 40`

OlivierSohn replied:
good catch!

mpilgrem commented on System/Console/ANSI/Windows/Emulator.hs:
```diff
@@ -294,21 +293,31 @@ applyANSISGRToAttribute def sgr attribute = case sgr of
   SetColor Background Vivid color -> applyBackgroundANSIColorToAttribute color
     (attribute .|. bACKGROUND_INTENSITY)
+  SetColor8 layer color -> applyANSISGRToAttribute def (SetColor8Code layer $ color8ToCode color) attribute
+  SetColor8Code layer code -> let intensity = fst res
```
I wondered if:
~~~~
let (intensity, col) = approximateColor8Code code
in  applyANSISGRToAttribute def (SetColor layer intensity col) attribute
~~~~
would be 'clearer' code?

OlivierSohn replied:
oh yes, indeed

OlivierSohn commented:
Thanks for the review @mpilgrem, if you have no further comments I'll squash to make it "mergeable".

OlivierSohn commented:
Hello, I force-pushed the squashed commit, PTAL

feuerbach commented on System/Console/ANSI/Codes.hs:
```diff
@@ -92,6 +94,17 @@ colorToCode color = case color of
+-- | 'color8ToCode' @color@ returns the index of the color
+color8ToCode :: Color8 -> Color8Code
+color8ToCode (Gray8Color g)       = Color8Code $ 232 + clamp g 0 23
+color8ToCode (RGB8Color r' g' b') = Color8Code $ 16 + 36 * r + 6 * g + b
```
Please add an explanation or reference for how/why this works

OlivierSohn replied:
I'll add some comments, it works because `0x10-0xE7:  6 × 6 × 6 cube (216 colors): 16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5)` (from https://en.wikipedia.org/wiki/ANSI_escape_code)

feuerbach commented on System/Console/ANSI/Color.hs:
```diff
@@ -0,0 +1,102 @@
+
+color8To24 :: Color8 -> SRGBLin.Colour Float
+color8To24 (RGB8Color r' g' b') =
+  let mapping = (\v -> v/255.0) . fromIntegral . xtermMapRGB8bitComponent . (\v -> clamp v 0 5) :: Int -> Float
```
Some explanation would be nice here, too.

OlivierSohn replied:
which aspect exactly would you like to see more documented? the clamp or the division by 255, or both?

OlivierSohn replied:
btw the clamp is also a consequence of `0x10-0xE7: 6 × 6 × 6 cube (216 colors): 16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5)`, maybe I can mention that in the Module doc to avoid repeating the same doc at several places...

feuerbach commented on System/Console/ANSI/Color.hs:
```diff
@@ -0,0 +1,102 @@
+  in  SRGBLin.rgb g g g
+
+-- | how xterm interprets 8bit rgb colors (deduced from https://jonasjacek.github.io/colors/)
+{-# INLINE xtermMapRGB8bitComponent #-}
```
Are these INLINE pragmas needed?

OlivierSohn replied:
I was under the impression that these functions are "small enough" so that inlining them would make the program performance better.

feuerbach replied:
They are small enough, and ghc is probably smart enough to make them inlinable. And if ghc doesn't inline them, that perhaps it has a good reason to. So I wouldn't add INLINE pragmas unless there's  a good explicit reason to do so.

OlivierSohn replied:
ok, I removed them.

feuerbach commented on System/Console/ANSI/Types.hs:
```diff
@@ -14,7 +18,7 @@ import Data.Ix (Ix)
 import Data.Colour (Colour)
 
--- | ANSI colors: come in various intensities, which are controlled by
+-- | ANSI 3/4-bit "candy" colors: come in various intensities, which are controlled by
```
It's not obvious what "candy" means here, could you clarify?

OlivierSohn replied:
It's a term I saw here and there, for example in https://stackoverflow.com/questions/15682537/ansi-color-specific-rgb-sequence-bash

I can remove it, as it doesn't seem to be official and use "4-bit colors" instead

feuerbach replied:
Yeah, you can remove it or explain it — either way is fine with me.

OlivierSohn commented:
@feuerbach Thanks for the review, I added some more comments so PTAL. If you think INLINE pragmas could hurt I can remove them but I thought it could make the code faster. I might be wrong, though.

OlivierSohn commented:
@feuerbach I think I implemented all the changes, please tell me if it looks good to you, so that I squash then the commits.

mpilgrem commented:
@OlivierSohn and @feuerbach, having done some background reading about the standards behind `ESC [ 3 8 ; 5 ; <palette-code>; m`, I have been re-thinking on 'types'. I wonder if:

* the only additional constructor for type `SGR` should be something like `SetPaletteColor ConsoleLayer Color8Code` (because that is the only new SGR functionality that is being reflected)

* `Color8Code` should be type synonym for, or `newtype` wrapper around, `Word8` (rather than `Int`) (because this is a about reflecting an 8-bit palette only)

* utility functions should map from other colour types to `Color8Code` for the xterm 256 colour protocol - something like `toXtermPalette :: Colour a -> Color8Code` or (for monochrome, where `Y` is a suitable type synonym or newtype for luma values) `lumaToXtermPalette :: Y -> Color8Code` - recognising that the xterm 256 colour protocol is only one implementation of a 256-colour palette, albeit ubiquitous. I also wonder, is there really a need to introduce a new, specifically 6x6x6, RGB colour type (especially in the `ansi-terminal` package, which is not a 'colour' package as such) - as opposed to providing a mapping from `Colour a` to `Color8Code`?

* in respect of a function like `lumaToXtermPalette`, why limit it to the last 24 greys in the xterm 256-colour palette. That is, there are other greys in the palette (and black and white, of course).

I apologise that these thoughts are a little late in the day; I was busy the past week.

OlivierSohn commented:
Hello @mpilgrem I think these remarks make a lot of sense, however on my side I won't have much time anymore to allocate to implement them. But someone can continue based on this MR and your remarks of course.

mpilgrem commented:
@OlivierSohn, in that case, I will offer a pull request to your `feature/support-all-8bit-colors` branch on `OlivierSohn/ansi-terminal` to reflect what I have in mind.

OlivierSohn commented:
Hello, I think it is in a mergeable state, with the contribution of @mpilgrem on types, and I reworked the haddock documentation for Color, since it is now part of the public API. PTAL.

OlivierSohn commented:
Hello @feuerbach, PTAL, as I said in the previous comment I think it is in a mergeable state. (This is a duplicate of the previous message, sorry for spamming! But I'm not sure if you receive any notification if your name is not mentionned in the comment?)

feuerbach commented:
Hi @OlivierSohn, sorry, I wasn't following your discussion with Mike, so I'll leave this up to him to review.

OlivierSohn commented:
Ok, @mpilgrem PTAL (it's mostly your changes so you won't be surprised :) plus the COlor module which I exposed in the public API)

mpilgrem commented on System/Console/ANSI/Color.hs:
```diff
@@ -26,15 +31,18 @@ import System.Console.ANSI.Types
 
 -- | Returns a color, in the sRGB colorspace, whose components are the
 -- components of an 8-bit ANSI color.
```
I do not think the 'ANSI' standard specifies that the paletted colour should follow the xterm 256 colour palette protocol. On that basis, perhaps something like this would be a better description: 'Returns the color in the sRGB colorspace corresponding to the color in the xterm 256-color palette.'

mpilgrem commented on System/Console/ANSI/Codes.hs:
```diff
@@ -60,13 +60,18 @@ module System.Console.ANSI.Codes
     -- * Utilities
-  , colorToCode, csi, sgrToCode
+  , colorToCode
+  , color8CodeToXterm256
```
I am wondering if this module needs to export `color8CodeToXterm256`, `xterm256ColorToCode` or `closestXterm256Color`, if module `System.Console.ANSI.Color` is to be exposed.

OlivierSohn replied:
yes, these can be removed

mpilgrem commented on System/Console/ANSI/Codes.hs:
```diff
@@ -60,13 +60,18 @@ module System.Console.ANSI.Codes
 
     -- * Utilities
-  , colorToCode, csi, sgrToCode
+  , colorToCode
```
This may be an @feuerbach level decision: I am also wondering (see below) if this module should still export `colorToCode` if module `System.Console.ANSI.Color` is to export that function and be exposed. That would be a breaking change from the past.

OlivierSohn replied:
@mpilgrem @feuerbach, I'll leave it as-is for now, since it's a breaking change. I'll just add a comment saying that in the future major release, this should be removed.

OlivierSohn commented:
@mpilgrem I've changed the documentation, and to not make a breaking change I left the reexport of 'colorToCode' in Codes module. If needed I can remove it (cc @feuerbach)

mpilgrem commented:
@OlivierSohn, from my perspective, this can be squashed into a single logical commit. From past experience, I know that @feuerbach would prefer a comprehensive commit message too. I've tested it on Windows 10 (version 10.0.16299.64), including emulating emulation by turning off the test for ANSI-capable.

OlivierSohn commented:
@mpilgrem I squashed it, also I tested on OSX 10.13.1, it works fine (colors and grayscale).

OlivierSohn commented:
@feuerbach your review still appears with a red cross and a "requested changes" label, are there still open questions on your side?

feuerbach commented:
Hey guys,

Sorry to drag this, but I've found the time to give this a bit of attention, and I am not completely satisfied with the API.

Specifically, I don't think that the decision to have `Color8Code` as a field in `SetPaletteColor` is justified.
Other SGR commands are relatively high-level and can be understood by an average user, but this one for some reason is based on the ANSI code representation.

What I suggest is to substitute `Color8Code` with `Xterm256Color` itself. Most users do not need to know about the details of the protocol; they want to do some colorful output. Having an intermediary `Color8Code` type means that the user will spend more time figuring out how to use `SetPaletteColor`, and their code will contain an unnecessary `xterm256ColorToCode` call. Instead this conversion should happen in `sgrToCode`. Oliver, I think you said that you have a use case for a raw color code; in that case, it could be made another constructor of `Xterm256Color`.

Sorry if some of this goes against Mike's opinion (which I appreciate and value) and if you need to revert some of the changes you made earlier because of that.

Also, the documentation of `Xterm256Color`'s constructors is very cryptic. I don't see why the user of this library needs to know about the ANSI ranges; so unless you disagree, those docs should be turned into code comments. (They could remain where they are; just remove the haddock markers.) What the docs should clearly say but don't is what range of parameters each constructor takes (e.g. 0..23 for grayscale, 0..5 for RGB). It can be kinda inferred from the current docs, but it must be very clear and simple.

I can mistunderstand something or just be wrong — please tell me if so.

feuerbach commented:
Also, the docs for `SystemColor` should probably mention that `SetPaletteColor` with it has the same meaning/effect as `SetColor`, to avoid confusion.

OlivierSohn commented:
@feuerbach I think what you propose makes sense. Indeed, I have a use case where I want to give a raw color8code. I won't be able however to refactor anything in the coming days, so if someone wants to do it, feel free. you can push commits on this branch I think.

OlivierSohn commented:
Thinking about it, I don't think adding the raw Color8Code constructor to XTerm256Color would be practical, as it leads to 2 different ways of representing the same thing, which is imho not good. We have a function that converts from a Color8Code to XTerm256Color, it should be enough for all use cases (I think if someone is able to understand ANSI 8but ranges, he or she won't mind calling a conversion function before passing the argument to setSGRCode)

feuerbach commented:
> which is imho not good

Why is it not good? From my point of view, XTerm256Color just lists all different ways you could specify a color.

> it should be enough for all use cases

It definitely is enough — I'm not saying it misses anything. I'm just saying it makes the users deal with concepts and functions that are unnecessary for their task.

> I think if someone is able to understand ANSI 8bit ranges, he or she won't mind calling a conversion function before passing the argument to setSGRCode

And that's exactly my problem — it should not be necessary to understand ANSI 8bit ranges to produce colorful output.

OlivierSohn commented:
@feuerbach  I agree that we should support directly XTerm256Color with SetPaletteColor, for the reasons you indicated.

But if Color8Code becomes a constructor of XTerm256Color, every function that takes a XTerm256Color will have an additional pattern match for Color8Code where it will convert the Color8Code to a RGBColor or GrayColor of SystemColor and then call the function again on that. It seems very convoluted, and sign of "something is wrong" to me. Don't you agree?

feuerbach commented:
> But if Color8Code becomes a constructor of XTerm256Color, every function that takes a XTerm256Color will have an additional pattern match for Color8Code where it will convert the Color8Code to a RGBColor or GrayColor or SystemColor and then call the function again on that.

What functions do you have in mind? I thought the only purpose of XTerm256Color is to be converted to Color8Code.

OlivierSohn commented:
Many :) look at Color.hs, a lot of them use this type

feuerbach commented:
Ah, I see, thanks. Let me think about it.

feuerbach commented:
Alright, in that case, how about only using XTerm256Color in the SGR constructor?

Then the easy things are easy and hard things (using the 8 bit color code, as in your case) are possible (with color8CodeToXterm256)?

feuerbach commented:
> In my case, I use raw color codes because I do a lot of comparisons between colors, I need to store them encoded in a vector, etc.. so I prefer a small type, as it leads to faster code and better cache usage.

>  Also, in theory, keeping Color8Code could also make GHC see more inlining opportunities for those who use it.

Interesting, but you didn't really explain what your use case is, and I still have no idea how often people care about the performance of this package (I'd say they mostly don't). I don't mind improving efficiency when it's free (like in #42), but I'd rather not make the API more complicated for 99% of users in order to accommodate one unusual use case.

OlivierSohn commented:
I see.

If you want to know more about my use case, I invite you to take a look at the haddoc doc of 
https://github.com/OlivierSohn/hamazed/blob/master/src/Render/Backends/Delta.hs

feuerbach commented:
Looks very cool! But it confirms my suspicion that this is a rather unusual use case.

OlivierSohn commented:
Ok, I understand your point. I'll leave this branch as-is, and anyone can continue to finish it according to what you want for the API.

mpilgrem commented:
@feuerbach, I've picked up the suggestion @OlivierSohn made, as a new pull request #45, hopefully reflecting the comments that you made about the desired API design.

mpilgrem commented:
Closed as resolved by #45.
